File: LC33.py
class Solution(object):
    def search(self, nums, k):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        low = 0
        high = len(nums)-1
        while low<=high:
            mid = (low+high)//2
            if nums[mid] == k:
                return mid
            if nums[mid]<=nums[high]: #lright part is sorted
                if nums[mid] <= k and nums[high] >= k:
                    low = mid+1
                else:
                    high = mid-1
            else:
                if nums[low]<=k and nums[mid]>=k:
                    high=mid-1
                else:
                    low=mid+1
        return -1

File: LC33.java
// 33. Search in Rotated Sorted Array

// TC: O(log(n))
// SC: O(1)
// Solved in LC: Yes

// Approach 1
class Solution {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0)
            return -1;
        int n = nums.length;
        int l = 0, h = n - 1;
        while (l <= h) {
            int mid = (l + h) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[l] <= nums[mid]) {
                if (target >= nums[l] && target < nums[mid])
                    h = mid - 1;
                else
                    l = mid + 1;
            } else {
                if (target > nums[mid] && target <= nums[h])
                    l = mid + 1;
                else
                    h = mid - 1;
            }
        }
        return -1;
    }
}


// Approach 2
class Solution2 {
    public int binarySearch(int[] nums, int target, int l, int h) {
        while (l <= h) {
            int mid = (l + h) / 2;
            if (nums[mid] == target)
                return mid;
            else if (nums[mid] > target)
                h = mid - 1;
            else
                l = mid + 1;
        }
        return -1;
    }

    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0)
            return -1;
        int len = nums.length;
        int l = 0, h = len - 1;
        int ref = nums[0];
        while (l <= h) {
            int mid = (l + h) / 2;
            if (nums[mid] < ref)
                h = mid - 1;
            else
                l = mid + 1;
        }
        System.out.println(l + " " + h);
        int left = binarySearch(nums, target, 0, h);
        if (left >= 0)
            return left;
        return binarySearch(nums, target, l, len - 1);
    }
}